make target det list export

walterddr opened the pull request:
Add functionality to export target det list.

Currently it outputs a list of test files that TD decided to run. What extra information do we want to export?
- [ ] file type? 
- [ ] ALL test files that ran through TD and a flag true/false? (b/c some files are not even going through TD, it might be good to log everything)
- [ ] ALL test files dependencies that triggered TD to decide to run (e.g. which traceback changed files triggered the decision of running a test)

Commit:
adding functionality to export target determination list

also add gitignore

Apply suggestions from code review

fix mypy

Update tools/testing/test_selections.py

walterddr commented on test/run_test.py:
```diff
@@ -1051,6 +1061,13 @@ def main():
+        # TODO: move this to tools/test_selections.py
+        target_det_filename = options.export_target_det_list
+        if target_det_filename:
+            print(f'Determine target list and report.')
```
```suggestion
            print(f'Determine target list and report to {target_det_filename}.')
```

walterddr commented on tools/testing/test_selections.py:
```diff
@@ -290,6 +290,17 @@ def export_S3_test_times(test_times_filename: Optional[str] = None) -> Dict[str,
 
 
+# TODO refactor this and unify with tools.stats.export_*
+def export_target_det_list(selected_test, target_det_filename: str) -> Dict[str, float]:
```
```suggestion
def export_target_det_list(selected_test: List[Any], target_det_filename: str) -> Dict[str, float]:
```

walterddr commented on tools/testing/test_selections.py:
```diff
@@ -290,6 +290,17 @@ def export_S3_test_times(test_times_filename: Optional[str] = None) -> Dict[str,
 
 
+# TODO refactor this and unify with tools.stats.export_*
+def export_target_det_list(selected_test: List[Any], target_det_filename: str) -> Dict[str, float]:
```
```suggestion
def export_target_det_list(selected_test: List[Any], target_det_filename: str) -> None:
```

zhouzhuojie commented on tools/testing/test_selections.py:
```diff
@@ -290,6 +290,17 @@ def export_S3_test_times(test_times_filename: Optional[str] = None) -> Dict[str,
+        print(f'Exporting target determination result to {target_det_filename}.')
+        if os.path.exists(target_det_filename):
+            print(f'Overwriting existent file: {target_det_filename}')
+        with open(target_det_filename, 'w+') as file:
+            json.dump(selected_test, file, indent='    ', separators=(',', ': '))
+            file.write('\n')
```
do we export it to artifact or other place, wondering how can we read the file content once saved

walterddr replied:
this is only meant for local use. At this point, this PR only provides this capability to export to a file (similar to export_slow_test), it does not enable using it on CI, which will be done in a separate PR.

Commit:
adding more funtionality
1. record TD results and distinguish whether it is skipped or it is determined as enabled
2. also record reason as file name and file type that triggered it
